lib/scholar/metrics.ex

```diff
@@ -336,6 +336,92 @@ defmodule Scholar.Metrics do
     Nx.indexed_add(zeros, indices, updates)
   end
 
+  @doc ~S"""
+  Calculates F1 score given rank-1 tensors which represent
+  the expected (`y_true`) and predicted (`y_pred`) classes.
+
+  ## Options
+
+    * `:num_classes` - required. Number of classes contained in the input tensors
+    * `:average` - optional. This determines the type of averaging performed on the data.
+      * `:macro`. Calculate metrics for each label, and find their unweighted mean.
+        This does not take label imbalance into account.
+      * `:weighted`. Calculate metrics for each label, and find their average weighted by
+        support (the number of true instances for each label).
+      * `:micro`. Calculate metrics globally by counting the total true positives,
+        false negatives and false positives.
+      * If not specified, then the f1 scores for each class are returned.
+
+  ## Examples
+
+      iex> y_true = Nx.tensor([0, 1, 1, 1, 1, 0, 2, 1, 0, 1], type: {:u, 32})
+      iex> y_pred = Nx.tensor([0, 2, 1, 1, 2, 2, 2, 0, 0, 1], type: {:u, 32})
+      iex> Scholar.Metrics.f1_score(y_true, y_pred, num_classes: 3, average: nil)
+      #Nx.Tensor<
+        f32[3]
+        [0.6666666865348816, 0.6666666865348816, 0.4000000059604645]
+      >
+      iex> Scholar.Metrics.f1_score(y_true, y_pred, num_classes: 3, average: :macro)
+      #Nx.Tensor<
+        f32
+        0.5777778029441833
+      >
+      iex> Scholar.Metrics.f1_score(y_true, y_pred, num_classes: 3, average: :weighted)
+      #Nx.Tensor<
+        f32
+        0.64000004529953
+      >
+      iex> Scholar.Metrics.f1_score(y_true, y_pred, num_classes: 3, average: :micro)
+      #Nx.Tensor<
+        f32
+        0.6000000238418579
+      >
+  """
+  defn f1_score(y_true, y_pred, opts \\ []) do
+    opts = keyword!(opts, [:num_classes, :average])
+
+    assert_shape_pattern(y_true, {_})
+    assert_shape(y_pred, Nx.shape(y_true))
+
+    transform(opts[:average], fn average ->
+      if Elixir.Kernel.==(average, :micro) do
+        accuracy(y_true, y_pred)
+      else
+        num_classes =
+          transform(opts[:num_classes], fn num_classes ->
+            num_classes || raise ArgumentError, "missing option :num_classes"
+          end)
```

Comment:
I think it makes sense to require it for consistency, so let's move it to the top.

Reply:
sure!